File: python/app_pkg/domains/reporting.py
"""reporting — a self-contained, owner-scoped read-side AGGREGATION store (a CQRS read model the app FEEDS; it does
NOT read other domains' stores — the boundary rule). A caller ingests typed FACT rows into a named dataset, then POSTs
a GROUP BY rollup (COUNT/SUM/MIN/MAX + equality filters + deterministic order + bounded pagination), and can drain a
dataset. Dangerous properties, all proven ×3 (python/go/node):
(1) AGGREGATION CORRECTNESS: COUNT/SUM/MIN/MAX are EXACTLY right and byte-identical ×3 (the search retrieval-honesty
    analog — a wrong sum still returns 200); the invariant recomputes the expectation independently.
(2) OWNER-SCOPED AGGREGATION: an aggregate NEVER includes another owner's facts. The owner is the bearer subject
    (require_identity, NEVER a client field) and is a MANDATORY conjunct on every scan (query/list/drain); a stranger
    gets an empty result, never a 403 (existence never leaks).
(3) DERIVED-OVERFLOW SAFETY: measure inputs are STRICT integers bounded ±(2^53-1) (reject 5.0/5.5/str/bool, ×3); the
    SUM is a derived value accumulated over a scan — it PREDICTS overflow BEFORE each add and fails loud (422) at
    2^53, so no wrapped go-int64 / precision-lost node-float is ever returned.
(4) DETERMINISTIC GROUP ORDER: groups are keyed + ordered by the digest_hex of their PRE-HASHED group-by values
    (an ASCII-hex key — injective and byte-identical ×3, sidestepping the python-codepoint / node-UTF-16 sort split),
    so the same query returns groups in the same order in all three languages regardless of store scan order.
Exactly-once ingest: a fact is immutable and idempotent on (owner, dataset, key) via scoped_key + claim_once (a
re-POST returns the SAME fact; corrections are compensating events — see INTEROP.md). Every route require_identity;
durable across restart. The in-process store.values scan is the documented store-swap-at-scale limit (see INTEROP.md)."""
from typing import List, Optional

from fastapi import APIRouter, Depends, Request
from pydantic import BaseModel, Field, StrictStr

from ..core import clock, store
from ..core.errors import invalid, require_identity
from ..parts.digest import digest_hex, scoped_key
from ..parts.idempotent_claim import claim_once
from ..parts.paginate import paginate
from ..parts.well_formed import make_well_formed, require_well_formed

router = APIRouter(prefix="/reporting", tags=["reporting"])
_ROUTE = "POST /reporting/facts"          # the owner-scoped fact-slot discriminator (same string ×3, used by the drain)
_MAX = 9007199254740991                   # 2**53-1: the magnitude every language holds EXACTLY (the ×3-safe ceiling)
_OPS = ("count", "sum", "min", "max")

# state in store: seq "reporting_fact" the monotonic id · ns "reporting_facts"
# scoped_key(route, owner, digest_hex(_h(dataset), _h(key))) -> the WHOLE record
# {id, owner, dataset, key, dimensions:{str:str}, measures:{str:int}, created_at} in ONE atomic claim. The slot is
# OWNER-scoped (scoped_key hashes the owner IN) AND injective in (dataset,key) (each pre-hashed). Same names ×3.


def _h(s: str) -> str:
    # pre-hash a component so a ':'-containing value can't forge another key — digest_hex joins with ':', and a
    # pre-hashed component is fixed-width colon-free hex (the relations injective-preimage idiom).
    return digest_hex(s)


def _clean(s: str, what: str) -> str:
    # contain-BEFORE-hash then validate: a lone surrogate -> U+FFFD (so a stored value never 5xxs a later re-read),
    # THEN reject empty / control (<0x20) / >1024 code points. Applied to EVERY response-bound string incl. map KEYS
    # and the query's group_by / aggregate `as` names (an un-contained key would be a stored 5xx poison).
    return require_well_formed(make_well_formed(s), what)


def _strict_int(v, what: str) -> int:
    # a JSON INTEGER in the ×3-safe range: reject a float (5.0/5.5 -> not int ×3), a bool (a distinct JSON type),
    # a string, and a magnitude past ±(2^53-1). `type(v) is int` excludes bool (type(True) is bool) AND float.
    if type(v) is not int or abs(v) > _MAX:
        raise invalid(what)
    return v


class FactIn(BaseModel):
    dataset: StrictStr
    key: StrictStr
    dimensions: dict = {}                  # {str: str} — grouped/filtered on; validated below (no owner/id field exists -> no mass-assignment)
    measures: dict = {}                    # {str: strict-int} — summed; validated below


class AggIn(BaseModel):
    op: str
    field: Optional[str] = None
    as_: Optional[str] = Field(default=None, alias="as")   # 'as' is a python keyword -> alias; the output name


class QueryIn(BaseModel):
    dataset: StrictStr
    group_by: List[str] = []               # dimension names (list[str] -> a non-string element is 422)
    aggregate: List[AggIn] = []
    filter: dict = {}                      # {dim: value} equality, implicit AND


def _public(rec: dict) -> dict:
    return {"id": rec["id"], "owner": rec["owner"], "dataset": rec["dataset"], "key": rec["key"],
            "dimensions": rec["dimensions"], "measures": rec["measures"], "created_at": rec["created_at"]}


@router.post("/facts", status_code=201)
def create(data: FactIn, request: Request, owner: str = Depends(require_identity)) -> dict:
    dataset = _clean(data.dataset, "dataset must be non-empty with no control characters")
    key = _clean(data.key, "key must be non-empty with no control characters")
    if not isinstance(data.dimensions, dict) or not isinstance(data.measures, dict):
        raise invalid("dimensions and measures must be objects")
    dims = {}
    for k, v in data.dimensions.items():
        if not isinstance(v, str):
            raise invalid("dimension values must be strings")
        dims[_clean(k, "dimension name")] = _clean(v, "dimension value")   # contain KEY and value (contain-before-store)
    meas = {}
    for k, v in data.measures.items():
        meas[_clean(k, "measure name")] = _strict_int(v, "measure values must be integers in the safe range")
    now = clock.current(request)
    slot = scoped_key(_ROUTE, owner, digest_hex(_h(dataset), _h(key)))     # owner-scoped + injective in (dataset,key); it IS the id
    prior = store.get("reporting_facts", slot)                            # fast path: a settled fact never re-writes
    if prior is None:
        rec = {"id": slot, "owner": owner, "dataset": dataset, "key": key,
               "dimensions": dims, "measures": meas, "created_at": now}   # id = the deterministic slot; owner from the token
        prior = claim_once("reporting_facts", slot, rec)                  # exactly-once: a repeat (dataset,key) returns the winner (immutable)
    return _public(prior)


@router.get("/facts")
def list_facts(owner: str = Depends(require_identity), dataset: str = "", limit: str = "", cursor: str = "") -> dict:
    # SCOPED read: only the caller's own facts leave the store (filtered on the authenticated owner FIELD as stored,
    # never a client value), id-sorted for a stable paged walk, then a BOUNDED page; a stranger gets an empty
    # page, never 403.
    ds = make_well_formed(dataset)                                        # contain the optional filter (empty = no filter)
    mine = sorted((_public(r) for r in store.values("reporting_facts")
                   if r["owner"] == owner and (ds == "" or r["dataset"] == ds)), key=lambda r: r["id"])
    page, nxt, ok = paginate(mine, cursor, limit)
    if not ok:
        raise invalid("invalid cursor or limit")
    return {"results": page, "next_cursor": nxt}


def _validate_aggs(aggregate: List[AggIn]) -> list:
    aggs, seen = [], set()
    for a in aggregate:
        if a.op not in _OPS:
            raise invalid("unknown aggregate op")
        if a.op == "count":
            if a.field is not None:
                raise invalid("count takes no field")
            field, as_ = None, _clean(a.as_ if a.as_ is not None else "count", "aggregate name")
        else:
            if a.field is None:
                raise invalid(a.op + " requires a field")
            field = _clean(a.field, "aggregate field")
            as_ = _clean(a.as_ if a.as_ is not None else a.op + "_" + field, "aggregate name")
        if as_ in seen:
            raise invalid("duplicate aggregate name")
        seen.add(as_)
        aggs.append((a.op, field, as_))
    if not aggs:
        raise invalid("at least one aggregate is required")
    return aggs


@router.post("/query")
def query(data: QueryIn, owner: str = Depends(require_identity), limit: str = "", cursor: str = "") -> dict:
    # SCOPED aggregation: the owner conjunct is INLINE (never a shared helper — the owner check stays visible at the
    # read site);
    # the O(n) facts scan is the documented store-swap-at-scale limit, the GROUPS result rides paginate (unbounded
    # group cardinality).
    dataset = _clean(data.dataset, "dataset must be non-empty with no control characters")
    group_by = [_clean(n, "group_by name") for n in data.group_by]
    aggs = _validate_aggs(data.aggregate)
    if not isinstance(data.filter, dict):
        raise invalid("filter must be an object")
    filt = {}
    for k, v in data.filter.items():
        if not isinstance(v, str):
            raise invalid("filter values must be strings")
        filt[_clean(k, "filter name")] = _clean(v, "filter value")
    # id-sorted so the SUM accumulation order (hence any overflow trip) is deterministic ×3
    matching = sorted((r for r in store.values("reporting_facts")
                       if r["owner"] == owner and r["dataset"] == dataset
                       and all(r["dimensions"].get(fk) == fv for fk, fv in filt.items())),
                      key=lambda r: r["id"])
    groups: dict = {}
    for r in matching:
        values = [r["dimensions"].get(n) for n in group_by]              # None for a missing dimension
        kh = digest_hex(*[_h(v) if v is not None else "" for v in values])   # injective, ASCII-hex, ×3-identical
        g = groups.get(kh)
        if g is None:
            g = {"values": values, "count": 0, "sum": {}, "min": {}, "max": {}}
            groups[kh] = g
        g["count"] += 1
        for op, field, as_ in aggs:
            if op == "count" or field not in r["measures"]:
                continue
            v = r["measures"][field]
            if op == "sum":
                acc = g["sum"].get(as_, 0)
                if (v > 0 and acc > _MAX - v) or (v < 0 and acc < -_MAX - v):   # predict overflow BEFORE the add
                    raise invalid("an aggregate sum exceeds the safe integer range")
                g["sum"][as_] = acc + v
            elif op == "min":
                g["min"][as_] = v if as_ not in g["min"] else min(g["min"][as_], v)
            elif op == "max":
                g["max"][as_] = v if as_ not in g["max"] else max(g["max"][as_], v)
    out = []
    for kh in sorted(groups):                                            # ASCII-hex order, identical ×3
        g = groups[kh]
        key_obj = {n: val for n, val in zip(group_by, g["values"])}
        vals = {}
        for op, field, as_ in aggs:
            if op == "count":
                vals[as_] = g["count"]
            elif op == "sum":
                vals[as_] = g["sum"].get(as_, 0)                         # SUM of no matching values = 0 (documented)
            elif op == "min" and as_ in g["min"]:
                vals[as_] = g["min"][as_]                                # MIN/MAX of no values -> OMITTED (never 0)
            elif op == "max" and as_ in g["max"]:
                vals[as_] = g["max"][as_]
        out.append({"key": key_obj, "values": vals})
    page, nxt, ok = paginate(out, cursor, limit)
    if not ok:
        raise invalid("invalid cursor or limit")
    return {"groups": page, "next_cursor": nxt}


@router.delete("/facts")
def drain(request: Request, owner: str = Depends(require_identity)) -> dict:
    # OWNER-scoped filtered DRAIN (the relations bulk-delete precedent): dataset is REQUIRED (>=1 anchor — a bare
    # drain is 422, never delete-all-by-accident) + optional ?<dim>=<val> equality filters. unbounded-safe: a filtered
    # delete drains ALL matching (a bounded page would strand rows); owner conjunct INLINE, only the caller's facts.
    q = request.query_params
    ds_raw = q.get("dataset")
    if not ds_raw:
        raise invalid("dataset is required")
    dataset = _clean(ds_raw, "dataset must be non-empty with no control characters")
    filt = {}
    for k in q.keys():
        if k == "dataset":
            continue
        filt[_clean(k, "filter name")] = _clean(q.get(k), "filter value")
    deleted = 0
    for r in list(store.values("reporting_facts")):
        if r["owner"] == owner and r["dataset"] == dataset \
                and all(r["dimensions"].get(fk) == fv for fk, fv in filt.items()):
            store.delete_("reporting_facts", r["id"])                     # id IS the slot (deterministic scoped_key)
            deleted += 1
    return {"deleted": deleted}

File: node/src/domains/reporting.js
// reporting — a self-contained, owner-scoped read-side AGGREGATION store (a CQRS read model the app FEEDS; it does NOT
// read other domains' stores — the boundary rule). Matches python/go; durable. Dangerous properties: (1) AGGREGATION
// CORRECTNESS ×3 (COUNT/SUM/MIN/MAX byte-identical — a wrong sum still returns 200; the invariant recomputes it).
// (2) OWNER-SCOPED AGGREGATION — owner = requireIdentity (never a client field), a mandatory conjunct on every scan;
// a stranger gets an empty result, never 403. (3) DERIVED-OVERFLOW SAFETY — measures are STRICT ints bounded
// ±(2^53-1) (isStrictInt rejects 5.0/5.5/str ×3); the SUM predicts overflow BEFORE each add and fails loud (422) at
// 2^53 (no precision-lost float). (4) DETERMINISTIC GROUP ORDER — groups keyed+ordered by the digestHex of their
// PRE-HASHED group-by values (ASCII hex — injective + byte-identical ×3, dodging the codepoint/UTF-16 sort split).
// Exactly-once ingest via scopedKey+claimOnce (a fact is immutable). Every route requireIdentity. The in-process
// scan is the documented store-swap-at-scale limit (INTEROP.md).
import { isStrictInt, problem, requireIdentity, sendJSON, storeDelete, storeGet, storeValues, testNow } from '../core/runtime.js';
import { digestHex, scopedKey } from '../parts/digest.js';
import { claimOnce } from '../parts/idempotent_claim.js';
import { paginate } from '../parts/paginate.js';
import { isWellFormed, makeWellFormed } from '../parts/well_formed.js';

const ROUTE = 'POST /reporting/facts'; // the owner-scoped fact-slot discriminator (same string ×3)
const MAX = 9007199254740991;          // 2^53-1: the ×3-safe integer ceiling
const OPS = new Set(['count', 'sum', 'min', 'max']);

// state: seq "reporting_fact" · ns "reporting_facts" scopedKey(route, owner, digestHex(_h(dataset), _h(key))) -> the
// whole record {id, owner, dataset, key, dimensions:{str:str}, measures:{str:int}, created_at}. Owner-scoped + injective.

const _h = (s) => digestHex(s); // pre-hash a component (colon-free -> injective join)

// contain-BEFORE-hash then validate: surrogate -> U+FFFD (a stored value never 5xxs a re-read), THEN reject
// empty/control/>1024cp. Applied to EVERY response-bound string incl. map KEYS + group_by/`as` names. null = invalid.
function clean(s) {
  const c = makeWellFormed(s);
  return isWellFormed(c) ? c : null;
}

const publicView = (r) => ({ id: r.id, owner: r.owner, dataset: r.dataset, key: r.key,
  dimensions: r.dimensions, measures: r.measures, created_at: r.created_at });

export async function reportingFactsCreate(req, res, params, body) {
  const owner = await requireIdentity(req, res); // the runtime parsed the body already, so identity-first is fine
  if (owner === null) return;
  if (!body || typeof body.dataset !== 'string' || typeof body.key !== 'string') return problem(res, 422, 'invalid body');
  const dataset = clean(body.dataset);
  const key = clean(body.key);
  if (dataset === null || key === null) return problem(res, 422, 'dataset and key must be non-empty with no control characters');
  const dimsIn = body.dimensions === undefined ? {} : body.dimensions;
  const measIn = body.measures === undefined ? {} : body.measures;
  if (typeof dimsIn !== 'object' || dimsIn === null || Array.isArray(dimsIn)
      || typeof measIn !== 'object' || measIn === null || Array.isArray(measIn)) {
    return problem(res, 422, 'dimensions and measures must be objects');
  }
  const dims = {};
  for (const [k, v] of Object.entries(dimsIn)) {
    if (typeof v !== 'string') return problem(res, 422, 'dimension values must be strings');
    const ck = clean(k);
    const cv = clean(v);
    if (ck === null || cv === null) return problem(res, 422, 'dimension names and values must be non-empty with no control characters');
    dims[ck] = cv;
  }
  const meas = {};
  for (const k of Object.keys(measIn)) {
    if (!isStrictInt(measIn, k)) return problem(res, 422, 'measure values must be integers in the safe range'); // rejects 5.0/5.5/str/bool/>2^53 ×3
    const ck = clean(k);
    if (ck === null) return problem(res, 422, 'measure names must be non-empty with no control characters');
    meas[ck] = measIn[k];
  }
  const slot = scopedKey(ROUTE, owner, digestHex(_h(dataset), _h(key))); // owner-scoped + injective in (dataset,key)
  let prior = await storeGet('reporting_facts', slot);                   // fast path: a settled fact never re-writes
  if (prior === undefined) {
    const rec = { id: slot, owner, dataset, key, dimensions: dims, measures: meas, created_at: testNow(req) }; // id = the deterministic slot
    prior = await claimOnce('reporting_facts', slot, rec);              // exactly-once: a repeat (dataset,key) returns the winner
  }
  sendJSON(res, 201, publicView(prior));
}

export async function reportingFactsList(req, res) {
  const owner = await requireIdentity(req, res);
  if (owner === null) return;
  // SCOPED read: only the caller's own facts leave the store (owner FIELD filter, as stored — never a client value),
  // id-sorted, then a BOUNDED page; a stranger gets an empty page.
  const q = new URL(req.url, 'http://localhost').searchParams;
  const ds = makeWellFormed(q.get('dataset') || ''); // contain the optional filter (empty = no filter)
  const mine = (await storeValues('reporting_facts'))
    .filter((r) => r.owner === owner && (ds === '' || r.dataset === ds))
    .sort((a, b) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0)).map(publicView);
  const { items, next, ok } = paginate(mine, q.get('cursor') || '', q.get('limit') || '');
  if (!ok) return problem(res, 422, 'invalid cursor or limit');
  sendJSON(res, 200, { results: items, next_cursor: next });
}

// validateAggs -> { aggs } or { err }. Mirrors python/go.
function validateAggs(aggregate) {
  if (!Array.isArray(aggregate)) return { err: 'unknown aggregate op' };
  const aggs = [];
  const seen = new Set();
  for (const a of aggregate) {
    if (!a || typeof a !== 'object' || Array.isArray(a) || typeof a.op !== 'string' || !OPS.has(a.op)) return { err: 'unknown aggregate op' };
    let field = null;
    let as;
    if (a.op === 'count') {
      if (a.field !== undefined && a.field !== null) return { err: 'count takes no field' };
      if (a.as !== undefined && a.as !== null && typeof a.as !== 'string') return { err: 'aggregate name must be non-empty with no control characters' };
      as = clean(a.as !== undefined && a.as !== null ? a.as : 'count');
    } else {
      if (a.field === undefined || a.field === null || typeof a.field !== 'string') return { err: `${a.op} requires a field` };
      field = clean(a.field);
      if (field === null) return { err: 'aggregate field must be non-empty with no control characters' };
      if (a.as !== undefined && a.as !== null && typeof a.as !== 'string') return { err: 'aggregate name must be non-empty with no control characters' };
      as = clean(a.as !== undefined && a.as !== null ? a.as : `${a.op}_${field}`);
    }
    if (as === null) return { err: 'aggregate name must be non-empty with no control characters' };
    if (seen.has(as)) return { err: 'duplicate aggregate name' };
    seen.add(as);
    aggs.push({ op: a.op, field, as });
  }
  if (aggs.length === 0) return { err: 'at least one aggregate is required' };
  return { aggs };
}

export async function reportingQuery(req, res, params, body) {
  const owner = await requireIdentity(req, res);
  if (owner === null) return;
  // SCOPED aggregation: the owner conjunct is INLINE; the O(n) scan is the documented store-swap-at-scale limit, the
  // GROUPS result rides paginate.
  if (!body || typeof body.dataset !== 'string') return problem(res, 422, 'invalid body');
  const dataset = clean(body.dataset);
  if (dataset === null) return problem(res, 422, 'dataset must be non-empty with no control characters');
  const groupByIn = body.group_by === undefined ? [] : body.group_by;
  if (!Array.isArray(groupByIn)) return problem(res, 422, 'group_by must be a list');
  const groupBy = [];
  for (const n of groupByIn) {
    if (typeof n !== 'string') return problem(res, 422, 'group_by name must be non-empty with no control characters');
    const c = clean(n);
    if (c === null) return problem(res, 422, 'group_by name must be non-empty with no control characters');
    groupBy.push(c);
  }
  const va = validateAggs(body.aggregate === undefined ? [] : body.aggregate);
  if (va.err) return problem(res, 422, va.err);
  const aggs = va.aggs;
  const filterIn = body.filter === undefined ? {} : body.filter;
  if (typeof filterIn !== 'object' || filterIn === null || Array.isArray(filterIn)) return problem(res, 422, 'filter must be an object');
  const filt = {};
  for (const [k, v] of Object.entries(filterIn)) {
    if (typeof v !== 'string') return problem(res, 422, 'filter values must be strings');
    const ck = clean(k);
    const cv = clean(v);
    if (ck === null || cv === null) return problem(res, 422, 'filter names and values must be non-empty with no control characters');
    filt[ck] = cv;
  }
  const fEntries = Object.entries(filt);
  // id-sorted so the SUM accumulation order (hence any overflow trip) is deterministic ×3
  const matching = (await storeValues('reporting_facts'))
    .filter((r) => r.owner === owner && r.dataset === dataset && fEntries.every(([fk, fv]) => r.dimensions[fk] === fv))
    .sort((a, b) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0));
  const groups = new Map();
  for (const r of matching) {
    const values = groupBy.map((n) => (Object.hasOwn(r.dimensions, n) ? r.dimensions[n] : null));
    const kh = digestHex(...values.map((v) => (v !== null ? _h(v) : ''))); // injective, ASCII-hex, ×3-identical
    let g = groups.get(kh);
    if (g === undefined) { g = { values, count: 0, sum: {}, min: {}, max: {} }; groups.set(kh, g); }
    g.count += 1;
    for (const a of aggs) {
      if (a.op === 'count' || !Object.hasOwn(r.measures, a.field)) continue;
      const v = r.measures[a.field];
      if (a.op === 'sum') {
        const acc = a.as in g.sum ? g.sum[a.as] : 0;
        if ((v > 0 && acc > MAX - v) || (v < 0 && acc < -MAX - v)) return problem(res, 422, 'an aggregate sum exceeds the safe integer range');
        g.sum[a.as] = acc + v;
      } else if (a.op === 'min') {
        g.min[a.as] = a.as in g.min ? Math.min(g.min[a.as], v) : v;
      } else if (a.op === 'max') {
        g.max[a.as] = a.as in g.max ? Math.max(g.max[a.as], v) : v;
      }
    }
  }
  const out = [...groups.keys()].sort().map((kh) => { // default sort: lexicographic on ASCII hex -> identical ×3
    const g = groups.get(kh);
    const key = {};
    groupBy.forEach((n, i) => { key[n] = g.values[i]; });
    const vals = {};
    for (const a of aggs) {
      if (a.op === 'count') vals[a.as] = g.count;
      else if (a.op === 'sum') vals[a.as] = (a.as in g.sum ? g.sum[a.as] : 0); // SUM of no matching values = 0 (documented)
      else if (a.op === 'min') { if (a.as in g.min) vals[a.as] = g.min[a.as]; } // MIN/MAX of no values -> OMITTED
      else if (a.op === 'max') { if (a.as in g.max) vals[a.as] = g.max[a.as]; }
    }
    return { key, values: vals };
  });
  const q = new URL(req.url, 'http://localhost').searchParams;
  const { items, next, ok } = paginate(out, q.get('cursor') || '', q.get('limit') || '');
  if (!ok) return problem(res, 422, 'invalid cursor or limit');
  sendJSON(res, 200, { groups: items, next_cursor: next });
}

export async function reportingFactsDrain(req, res) {
  const owner = await requireIdentity(req, res);
  if (owner === null) return;
  // OWNER-scoped filtered DRAIN (the relations bulk-delete precedent): dataset REQUIRED (>=1 anchor) + optional
  // ?<dim>=<val> filters. unbounded-safe: a filtered delete drains ALL matching; owner conjunct INLINE.
  const q = new URL(req.url, 'http://localhost').searchParams;
  const dsRaw = q.get('dataset');
  if (!dsRaw) return problem(res, 422, 'dataset is required');
  const dataset = clean(dsRaw);
  if (dataset === null) return problem(res, 422, 'dataset must be non-empty with no control characters');
  const filt = {};
  for (const k of q.keys()) {
    if (k === 'dataset') continue;
    const ck = clean(k);
    const cv = clean(q.get(k));
    if (ck === null || cv === null) return problem(res, 422, 'filter names and values must be non-empty with no control characters');
    filt[ck] = cv;
  }
  const fEntries = Object.entries(filt);
  let deleted = 0;
  for (const r of await storeValues('reporting_facts')) {
    if (r.owner === owner && r.dataset === dataset && fEntries.every(([fk, fv]) => r.dimensions[fk] === fv)) {
      await storeDelete('reporting_facts', r.id); // id IS the slot (deterministic scoped_key)
      deleted += 1;
    }
  }
  sendJSON(res, 200, { deleted });
}
